plugins/python/python_plugin.c

```diff
@@ -1297,29 +1288,49 @@ void uwsgi_python_init_apps() {
 
 }
 
-void uwsgi_python_master_fixup(int step) {
 
-	static int master_fixed = 0;
-	static int worker_fixed = 0;
+void uwsgi_python_pre_uwsgi_fork() {
+	if (!uwsgi.master_process) return;
+
+	if (uwsgi.has_threads) {
+		// Since we are not calling cPython's fork, we need to avoid forking
+		// while the gil and import lock are held by a thread other than the
+		// main thread.
+		UWSGI_GET_GIL
+		_PyImport_AcquireLock();
+	}
+}
+
 
+void uwsgi_python_post_uwsgi_fork(int step) {
 	if (!uwsgi.master_process) return;
 
 	if (uwsgi.has_threads) {
 		if (step == 0) {
-			if (!master_fixed) {
-				UWSGI_RELEASE_GIL;
-				master_fixed = 1;
-			}
-		}	
+			// release locks witin master process
+			_PyImport_ReleaseLock();
+			UWSGI_RELEASE_GIL
+		}
 		else {
-			if (!worker_fixed) {
-				UWSGI_GET_GIL;
-				worker_fixed = 1;
-			}
+			// ensure thread state and locks is cleaned in child process
```

Comment:
```suggestion
			// ensure thread state and locks are cleaned in child process
```

Reply:
Any more information we can add here, like a link to `PyOS_AfterFork[_Child]` docs or source?

Why is this important to call, why should this not be removed?